src/rpc/index.js

```diff
@@ -0,0 +1,162 @@
+function convertHttpArgs (url, args) {
+  args = args || {}
+  const search = []
+  for (let k in args) {
+    if(is.string(args[k])) {
+      search.push(`${k}="${args[k]}"`)
+    } else if(Buffer.isBuffer(args[k])){
+      search.push(`${k}=0x${args[k].toString('hex')}`)
+    } else {
+      search.push(`${k}=${args[k]}`)
+    }
+  }
+  return `${url}?${search.join('&')}`
+}
+
+function convertWsArgs (args) {
+  args = args || {}
+  for (let k in args) {
+    let v = args[k]
+    if (typeof v === "number") {
+      args[k] = String(v)
+    } else if (Buffer.isBuffer(v)) {
+      args[k] = "0x" + v.toString("hex")
+    } else if (v instanceof Uint8Array) {
+      args[k] = "0x" + Buffer.from(v).toString("hex")
+    }
+  }
+  return args
+}
+
+const wsProtocols = [ "ws:", "wss:" ]
+const httpProtocols = [ "http:", "https:" ]
+const allProtocols = wsProtocols.concat(httpProtocols)
+
+class BaseRpc extends EventEmitter {
+  constructor (uriString = "localhost:27146") {
+    super()
+
+    // parse full-node URI
+    let { protocol, hostname, port } = url.parse(uriString)
+
+    // default to http
+    if (!allProtocols.includes(protocol)) {
+      let uri = url.parse(`http://${uriString}`)
+      protocol = uri.protocol
+      hostname = uri.hostname
+      port = uri.port
+    }
+
+    this.uri = !port ? `${protocol}//${hostname}/` : `${protocol}//${hostname}:${port}/`
+
+    if (wsProtocols.includes(protocol)) {
+      this.websocket = true
+      this.uri = `${this.uri}websocket`
+      this.call = this.callWs
+      this.connectWs()
+    } else if (httpProtocols.includes(protocol)) {
+      this.call = this.callHttp
+    }
+  }
+
+  connectWs () {
+    this.ws = pumpify(
+      ndjson.stringify(),
+      websocket(this.uri)
+    )
+
+    this.ws.on("error", (err) => this.emit("error", err))
+    this.ws.on("close", () => {
+      if (this.closed) return
+      this.emit("error", Error("websocket disconnected"))
+    })
+    this.ws.on("data", (data) => {
+      data = JSON.parse(data)
+      if (!data.id) return
+      this.emit(data.id, data.error, data.result)
+    })
+  }
+
+  callHttp (method, args) {
+    let url = this.uri + method
+    url = convertHttpArgs(url, args)
+    return axios({
+      url: url
+    }).then(function (res) {
+      // console.log(res)
+      const data = res.data
+      if (data.error) {
+        let err = Error(data.error.message)
+        Object.assign(err, data.error)
+        throw err
+      }
+      return data.result
+    }, function (err) {
+      throw Error(err)
+    })
+  }
+
+  callWs (method, args, listener) {
+    let self = this
+    return new Promise((resolve, reject) => {
+      let id = Math.random().toString(36)
+      let params = convertWsArgs(args)
+
+      if (method === "subscribe") {
+        if (typeof listener !== "function") {
+          throw Error("Must provide listener function")
+        }
+
+        // events get passed to listener
+        this.on(id + "#event", (err, res) => {
```

Comment:
If ws `throws`, this will not throw an error to the caller like `http` does. We might want to handle errors in the same way as `http`. During a call over `ws`, we can reject the promise with the error.

Reply:
only subscribe method can run here, so you have to add error event to your client which error can be catched there. if not subscribe but use client with ws, we will throw error back to caller
`// response goes to promise
        this.once(id, (err, res) => {
          if (err) return reject(err)
          resolve(res)
        })`

Reply:
What about [this error](https://github.com/binance-chain/javascript-sdk/blob/feature/rpc/src/rpc/index.js#L80) emitted by the WebSocket?

Reply:
will be cached in error event of the client